gateway-js/src/index.ts

```diff
@@ -551,28 +665,25 @@ export class ApolloGateway implements GraphQLService {
       this.logger.info('Updated Supergraph SDL was found.');
     }
 
-    await this.maybePerformServiceHealthCheck(result);
-
-    this.compositionId = result.id;
-    this.supergraphSdl = result.supergraphSdl;
+    this.compositionId = id;
+    this.supergraphSdl = supergraphSdl;
     this.parsedSupergraphSdl = parsedSupergraphSdl;
 
-    const { schema, supergraphSdl } = this.createSchemaFromSupergraphSdl(
-      result.supergraphSdl,
-    );
+    const { schema, supergraphSdl: generatedSupergraphSdl } =
+      this.createSchemaFromSupergraphSdl(supergraphSdl);
 
-    if (!supergraphSdl) {
+    if (!generatedSupergraphSdl) {
       this.logger.error(
         "A valid schema couldn't be composed. Falling back to previous schema.",
       );
     } else {
-      this.updateWithSchemaAndNotify(schema, supergraphSdl);
+      this.updateWithSchemaAndNotify(schema, generatedSupergraphSdl);
```

Comment:
There is still a re-entrancy concern in that one of the hooks called from here (or experimental_didUpdateComposition) could invoke the `update` hook again. I'm not sure if that would cause a problem or not.  I think it might make sense to have a little boolean state somewhere that externalSupergraphUpdateCallback sets at the beginning and clears at the end, and if it sees it's already set throws an error.

Reply:
Gotcha! I understand what you meant a lot better now. Should be resolved by c2a88ea